menas/ui/components/login/loginMain.controller.js

```diff
@@ -0,0 +1,137 @@
+    loginForm: {},
+
+    /**
+     * Called when a controller is instantiated and its View controls (if
+     * available) are already created. Can be used to modify the View before it
+     * is displayed, to bind event handlers and do other one-time
+     * initialization.
+     *
+     * @memberOf components.login.loginMain
+     */
+    onInit: function () {
+      this._router = sap.ui.core.UIComponent.getRouterFor(this);
+      let model = {username: "", password: "", submit: "Login"};
+
+      this.loginForm = this.byId("loginForm");
+      this.loginForm.setModel(new sap.ui.model.json.JSONModel(model), "login");
+
+      this.byId("loginSubmit").attachPress(this.onLoginSubmit, this);
+    },
+
+    onLoginSubmit: function (oEvent) {
+      let oData = this.loginForm.getModel("login").oData;
+
+      this._resetLoginFormState();
+      if (this._validateLogin(oData)) {
+        this._login(oData, this.byId("loginSubmit"))
+      }
+    },
+
+    _resetFieldState: function (sField) {
+      this.byId(sField).setValueState(sap.ui.core.ValueState.None);
+      this.byId(sField).setValueStateText("");
+    },
+
+    _resetLoginFormState: function () {
+      this._resetFieldState("username");
+      this._resetFieldState("password");
+    },
+
+    _validateField: function (oData, sField, sErrorMessage) {
+      let isOk = oData[sField] && oData[sField] !== "";
+
+      if (!isOk) {
+        let field = this.byId(sField);
+        field.setValueState(sap.ui.core.ValueState.Error);
+        field.setValueStateText(sErrorMessage);
+      }
+
+      return isOk;
+    },
+
+    _validateLogin(oData) {
+      let isValidUsername = this._validateField(oData, "username", "Username cannot be empty.");
+      let isValidPassword = this._validateField(oData, "password", "Password cannot be empty.");
+      return isValidUsername && isValidPassword;
+    },
+
+    _login: function (oData, oControl) {
+      if (oControl) oControl.setBusy(true);
+
+      let fnSuccess = (result, status, xhr) => {
+        let csrfToken = xhr.getResponseHeader("X-CSRF-TOKEN");
+        localStorage.setItem("csrfToken", csrfToken);
+        Functions.ajax("api/user/info", "GET", {}, (oInfo) => {
+          model.setProperty("/userInfo", oInfo);
+          this._router.navTo("schemas");
+        });
+      };
+
+      let fnError = () => {
+        MessageToast.show("Username or password incorrect");
+        this.byId("username").setValueState(sap.ui.core.ValueState.Error);
+        this.byId("password").setValueState(sap.ui.core.ValueState.Error);
+      };
+      $.ajax("api/login", {
+        complete: function () {
+          if (oControl) oControl.setBusy(false)
+        },
+        data: oData,
+        method: "POST",
+        success: fnSuccess,
+        error: fnError
+      })
+    },
+    /**
+     * Similar to onAfterRendering, but this hook is invoked before the controller's
+     * View is re-rendered (NOT before the first rendering! onInit() is used for
```

Comment:
Is this left here on purpose?

Reply:
Yeah, for consistency like in the other controllers. Maybe we can drop them, I don't know.